ksqldb-engine/src/main/java/io/confluent/ksql/planner/plan/JoinNode.java

```diff
@@ -214,14 +214,51 @@ protected int getPartitions(final KafkaTopicClient kafkaTopicClient) {
   @Override
   void validateKeyPresent(final SourceName sinkName, final Projection projection) {
 
-    final boolean atLeastOneKey = joinKey.getAllViableKeys(schema).stream()
-        .anyMatch(projection::containsExpression);
+    if (joinKey.isForeignKey()) {
+      final DataSource leftInputTable = getLeftmostSourceNode().getDataSource();
+      final SourceName leftInputTableName = leftInputTable.getName();
+      final String plainInputTableName = leftInputTableName.text();
+      final List<Column> leftInputTableKeys = leftInputTable.getSchema().key();
+      final List<Column> missingKeys =
+          leftInputTableKeys.stream().filter(
+              k -> !projection.containsExpression(
+                    new QualifiedColumnReferenceExp(leftInputTableName, k.name()))
+                  && !projection.containsExpression(new UnqualifiedColumnReferenceExp(
+                      ColumnName.of(plainInputTableName + "_" + k.name().text())
```

Comment:
This is duplicating the logic in `ColumnNames.generatedJoinColumnAlias()`, right? It'd be good to call the helper method to keep the logic in one place instead.

Reply:
It seems that generating an alias would generate new name (like prefix the name with `KSQL_COL`, but we what reconstruct on existing name, using the table name as prefix?

Reply:
No, if you look at the implementation for the method, you'll see that all it does is concatenate the source name with the column name, with an underscore in between. This is the same logic used throughout the rest of the code to rewrite join column names, which the logic here should be consistent with.